### PR TITLE
Update Mamba types and pass through use_cache attr to MambaModel

### PR DESCRIPTION
**Pull request by koayon**
# What does this PR do?

1. Previously the cache_params variable was inconsistently typed as `MambaCache`, `torch.Tensor` or `list[torch.Tensor]`. This PR updates this to `MambaCache` everywhere which is inline with the attributes that are being accessed in the logic. 
2. Previously if you ran the forward method with use_cache=True, the cache_params in the output would still be None as this argument wasn't being passed through to the `MambaModel`. This PR fixes this as below:
3. Also updates the docstrings in line with this change

Allowed the use_cache information to be passed through so that you can do: 
```python
import torch as t

from transformers import AutoTokenizer

if __name__ == "__main__":
    model = MambaForCausalLM(MambaConfig())
    tokeniser = AutoTokenizer.from_pretrained("state-spaces/mamba-130m-hf")
    input_ids: t.Tensor = tokeniser("Hey how are you doing?", return_tensors="pt")[  # type: ignore
        "input_ids"
    ]

    out: MambaCausalLMOutput = model(input_ids=input_ids, use_cache=True)
    assert out.cache_params is not None
    print(out.cache_params.ssm_states)
```

And get back the ssm_states, which was not previously possible

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?

@ArthurZucker
@gante

### CODE REVIEW

**Comment by amyeroberts** [comment]
Thanks for adding this @koayon 

Pinging @gante for first review of the cache logic, as @ArthurZucker is off this week

**Review by gante — APPROVED** [review]
LGTM, thank you for the PR 🤗 

I'd like a final check from @ArthurZucker, though -- there are some terminology updates in the docstrings, and I'm not very familiar with Mamba :)

File: src/transformers/models/mamba/modeling_mamba.py [hunk 1]
```diff
@@ -50,7 +50,13 @@
...
 is_fast_path_available = all(
-    (selective_state_update, selective_scan_fn, causal_conv1d_fn, causal_conv1d_update, mamba_inner_fn)
+    (
+        selective_state_update,
+        selective_scan_fn,
+        causal_conv1d_fn,
+        causal_conv1d_update,
+        mamba_inner_fn,
+    )
 )
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:60]: this is unrelated and is styling, should be reverted!

File: src/transformers/models/mamba/modeling_mamba.py [hunk 2]
```diff
@@ -268,34 +314,16 @@ def slow_forward(self, input_states, cache_params=None):
...
-
-        self.conv_states = {
-            i: torch.zeros(batch_size, intermediate_size, conv_kernel_size, device=device, dtype=dtype)
-            for i in range(config.num_hidden_layers)
-        }
-        self.ssm_states = {
-            i: torch.zeros(batch_size, intermediate_size, ssm_state_size, device=device, dtype=dtype)
-            for i in range(config.num_hidden_layers)
-        }
-
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:293]: if moved, let's just keep the styling of this one please

File: src/transformers/models/mamba/modeling_mamba.py [hunk 3]
```diff
@@ -150,18 +190,24 @@ def cuda_kernels_forward(self, hidden_states: torch.Tensor, cache_params=None):
...
                 )
 
             # 3. State Space Model sequence transformation
             # 3.a. input varying initialization of time_step, B and C
             ssm_parameters = self.x_proj(hidden_states.transpose(1, 2))
             time_step, B, C = torch.split(
-                ssm_parameters, [self.time_step_rank, self.ssm_state_size, self.ssm_state_size], dim=-1
+                ssm_parameters,
+                [self.time_step_rank, self.ssm_state_size, self.ssm_state_size],
+                dim=-1,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:210]: same here, unrelated change

File: src/transformers/models/mamba/modeling_mamba.py [hunk 4]
```diff
@@ -150,18 +190,24 @@ def cuda_kernels_forward(self, hidden_states: torch.Tensor, cache_params=None):
             else:
                 if cache_params is not None:
                     conv_states = nn.functional.pad(
-                        hidden_states, (self.conv_kernel_size - hidden_states.shape[-1], 0)
+                        hidden_states,
+                        (self.conv_kernel_size - hidden_states.shape[-1], 0),
                     )
                     cache_params.conv_states[self.layer_idx].copy_(conv_states)
                 hidden_states = causal_conv1d_fn(
-                    hidden_states, conv_weights, self.conv1d.bias, activation=self.activation
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:157]: snae gere

File: src/transformers/models/mamba/modeling_mamba.py [hunk 5]
```diff
@@ -150,18 +190,24 @@ def cuda_kernels_forward(self, hidden_states: torch.Tensor, cache_params=None):
             else:
                 if cache_params is not None:
                     conv_states = nn.functional.pad(
-                        hidden_states, (self.conv_kernel_size - hidden_states.shape[-1], 0)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:153]: same here, unrelated change

File: src/transformers/models/mamba/modeling_mamba.py [hunk 6]
```diff
@@ -91,7 +127,11 @@ def __init__(self, config, layer_idx):
         # projection of the input hidden states
         self.in_proj = nn.Linear(self.hidden_size, self.intermediate_size * 2, bias=config.use_bias)
         # selective projection used to make dt, B and C input dependant
-        self.x_proj = nn.Linear(self.intermediate_size, self.time_step_rank + self.ssm_state_size * 2, bias=False)
+        self.x_proj = nn.Linear(
+            self.intermediate_size,
+            self.time_step_rank + self.ssm_state_size * 2,
+            bias=False,
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:134]: same here

File: src/transformers/models/mamba/modeling_mamba.py [hunk 7]
```diff
@@ -541,7 +571,10 @@ def forward(
 
         if cache_params is None and use_cache:
             cache_params = MambaCache(
-                self.config, inputs_embeds.size(0), device=inputs_embeds.device, dtype=inputs_embeds.dtype
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:544]: unrelated change let's revert

File: src/transformers/models/mamba/modeling_mamba.py [hunk 8]
```diff
@@ -609,7 +642,12 @@ def _update_model_kwargs_for_generation(
         return model_kwargs
 
     def prepare_inputs_for_generation(
-        self, input_ids, cache_params=None, inputs_embeds=None, attention_mask=None, **kwargs
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:612]: same here

File: src/transformers/models/mamba/modeling_mamba.py [hunk 9]
```diff
@@ -653,6 +691,7 @@ def forward(
             inputs_embeds=inputs_embeds,
             output_hidden_states=output_hidden_states,
             return_dict=return_dict,
+            **kwargs,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:694]: why is this required? it should not. The cache params are passed right above

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the typing, there are a lot of unrelated changes, and I am not sure I understand how you got the `cache_params` to be None but adding kwargs to the call should not really be the solution ! 

`use_cache=True` returns the cache for me:




unless you are training.

File: src/transformers/models/mamba/modeling_mamba.py (hunk 9 above)
Comment (koayon) [reply src/transformers/models/mamba/modeling_mamba.py:694]: I believe it's the use_cache argument that needs to be passed in for this to work as expected - we could restrict to just passing that through?

File: src/transformers/models/mamba/modeling_mamba.py (hunk 1 above)
Comment (koayon) [reply src/transformers/models/mamba/modeling_mamba.py]: Thanks, I've updated the styling 👌

**Comment by koayon** [comment]
Hey @ArthurZucker, thanks for your review!
I've reverted the styling changes, thanks 🙏 

In terms of the image that you were sending, it's unfortunately not showing up for me. But without the change to pass in use_cache I don't see the cache_params being returned. If there's a difference for you, I've just thought that it might be how it's running on CUDA vs MPS/CPU?

I append the following to the file: 

```python
import torch as t

from transformers import AutoTokenizer

if __name__ == "__main__":
    model = MambaForCausalLM(MambaConfig())
    tokeniser = AutoTokenizer.from_pretrained("state-spaces/mamba-130m-hf")
    input_ids: t.Tensor = tokeniser("Hey how are you doing?", return_tensors="pt")["input_ids"]  # type: ignore

    out: MambaCausalLMOutput = model(input_ids=input_ids, use_cache=True)
    assert out.cache_params is not None
    print(out.cache_params.ssm_states)
 ```
 
 If the use_cache argument isn't passed through to the backbone (either with kwargs or separately as in the newer version), there is no cache_params returned and I get the error: 
 ```bash
python src/transformers/models/mamba/modeling_mamba.py      
The fast path is not available because on of `(selective_state_update, selective_scan_fn, causal_conv1d_fn, causal_conv1d_update, mamba_inner_fn)` is None. Falling back to the naive implementation. To install follow https://github.com/state-spaces/mamba/#installation and https://github.com/Dao-AILab/causal-conv1d
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Traceback (most recent call last):
  File "/[PATH_TO_TRANSFORMERS]/transformers/src/transformers/models/mamba/modeling_mamba.py", line 688, in <module>
    assert out.cache_params is not None
AssertionError
```

whereas with the use_cache argument being passed through I get a tensor returned: 
```bash
{0: tensor([[[-5.5237e-04,  9.6599e-04,  6.6771e-04,  ..., -5.3982e-04,
          -4.6061e-04, -7.0508e-04],
         [-3.7170e-05, -2.2089e-04, -1.0218e-04,  ...,  7.1232e-05, 
         ...
```

It does seem like this would be required for the expected behaviour.  Please let me know if you have any questions!

File: src/transformers/models/mamba/modeling_mamba.py (hunk 9 above)
Comment (koayon) [reply src/transformers/models/mamba/modeling_mamba.py]: Have amended this

File: src/transformers/models/mamba/modeling_mamba.py [hunk 10]
```diff
@@ -35,7 +35,6 @@
 from ...utils.import_utils import is_causal_conv1d_available, is_mamba_ssm_available
 from .configuration_mamba import MambaConfig
```
Comment (koayon) [inline src/transformers/models/mamba/modeling_mamba.py:37]:
```suggestion


```

File: src/transformers/models/mamba/modeling_mamba.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/models/mamba/modeling_mamba.py]: perfect

**Comment by ArthurZucker** [comment]
Alright, when using from_pretrained, the cache is used and passed subsequently, but not when using the initialization

File: src/transformers/models/mamba/modeling_mamba.py [hunk 11]
```diff
@@ -633,7 +635,7 @@ def forward(
         self,
         input_ids: Optional[torch.LongTensor] = None,
         inputs_embeds: Optional[torch.FloatTensor] = None,
-        cache_params: Optional[torch.FloatTensor] = None,
+        cache_params: Optional[MambaCache] = None,
         labels: Optional[torch.LongTensor] = None,
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:641]: let's add use_cache as an arg here

**Review by ArthurZucker — COMMENTED** [review]
Almost good to go!

File: src/transformers/models/mamba/modeling_mamba.py (hunk 11 above)
Comment (koayon) [inline src/transformers/models/mamba/modeling_mamba.py:641]:
```suggestion
        return_dict: Optional[bool] = None,
        use_cache: bool = False,
```

Comment (koayon) [reply src/transformers/models/mamba/modeling_mamba.py:641]: Added!

File: src/transformers/models/mamba/modeling_mamba.py [hunk 12]
```diff
@@ -633,10 +635,11 @@ def forward(
...
+        use_cache: bool = False,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:642]:
```suggestion
        use_cache: bool = None,
```

File: src/transformers/models/mamba/modeling_mamba.py [hunk 13]
```diff
@@ -647,12 +650,14 @@ def forward(
         """
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        use_cache = kwargs.get("use_cache", None)
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:653]:
```suggestion
use_cache = use_cache if use_cache is not None else config.use_cache
```
since you added use cache it is

**Review by ArthurZucker — COMMENTED** [review]
Sorry forgot about these!

**Comment by koayon** [comment]
@ArthurZucker great suggestion, didn't realise that was an attribute of the Config 👌 
Think it's ready to merge now 🚀

File: src/transformers/models/mamba/modeling_mamba.py (hunk 10 above)
Comment (koayon) [inline src/transformers/models/mamba/modeling_mamba.py:37]:
```suggestion


```

**Comment by ArthurZucker** [comment]
The failing test seems new, but it's because when training the use_cache should be disabled by the model

**Comment by ArthurZucker** [comment]
I'll have a look

File: src/transformers/models/mamba/modeling_mamba.py [hunk 14]
```diff
@@ -647,12 +650,14 @@ def forward(
         """
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:653]:
```suggestion
```
this should fix it!

Comment (koayon) [reply src/transformers/models/mamba/modeling_mamba.py:653]: Thanks! 🙏

**Review by ArthurZucker — APPROVED** [review]
Thanks for iterating!
